src/bindings.rs

```diff
@@ -0,0 +1,6192 @@
+                    }
+                }
+                impl ::std::cmp::Eq for IN_ADDR_0_1 {}
+                unsafe impl ::windows::Abi for IN_ADDR_0_1 {
+                    type Abi = Self;
+                    type DefaultType = Self;
+                }
+                pub type LPFN_ACCEPTEX =
+                    unsafe extern "system" fn(
+                        slistensocket: SOCKET,
+                        sacceptsocket: SOCKET,
+                        lpoutputbuffer: *mut ::std::ffi::c_void,
+                        dwreceivedatalength: u32,
+                        dwlocaladdresslength: u32,
+                        dwremoteaddresslength: u32,
+                        lpdwbytesreceived: *mut u32,
+                        lpoverlapped: *mut super::super::System::SystemServices::OVERLAPPED,
+                    )
+                        -> super::super::Foundation::BOOL;
+                pub type LPFN_CONNECTEX =
+                    unsafe extern "system" fn(
+                        s: SOCKET,
+                        name: *const SOCKADDR,
+                        namelen: i32,
+                        lpsendbuffer: *const ::std::ffi::c_void,
+                        dwsenddatalength: u32,
+                        lpdwbytessent: *mut u32,
+                        lpoverlapped: *mut super::super::System::SystemServices::OVERLAPPED,
+                    )
+                        -> super::super::Foundation::BOOL;
+                pub type LPFN_GETACCEPTEXSOCKADDRS = unsafe extern "system" fn(
+                    lpoutputbuffer: *const ::std::ffi::c_void,
+                    dwreceivedatalength: u32,
+                    dwlocaladdresslength: u32,
+                    dwremoteaddresslength: u32,
+                    localsockaddr: *mut *mut SOCKADDR,
+                    localsockaddrlength: *mut i32,
+                    remotesockaddr: *mut *mut SOCKADDR,
+                    remotesockaddrlength: *mut i32,
+                );
+                pub type LPWSAOVERLAPPED_COMPLETION_ROUTINE = unsafe extern "system" fn(
+                    dwerror: u32,
+                    cbtransferred: u32,
+                    lpoverlapped: *mut super::super::System::SystemServices::OVERLAPPED,
+                    dwflags: u32,
+                );
+                #[derive(:: std :: clone :: Clone, :: std :: marker :: Copy)]
+                #[repr(C)]
+                pub struct SCOPE_ID {
+                    pub Anonymous: SCOPE_ID_0,
+                }
+                impl SCOPE_ID {}
+                impl ::std::default::Default for SCOPE_ID {
+                    fn default() -> Self {
+                        unsafe { ::std::mem::zeroed() }
+                    }
+                }
+                impl ::std::cmp::PartialEq for SCOPE_ID {
+                    fn eq(&self, _other: &Self) -> bool {
+                        unimplemented!()
+                    }
+                }
```

Comment:
Good reason to treat `bindings.rs` as purely generated code since a fix for this is coming soon (implementing `PartialEq` for complex structs and unions) and you can simply rerun the `bindings` crate to pick up the improvements.

Reply:
> [...] you can simply rerun the `bindings` crate to pick up the improvements

That's what I mean with "the maintenance burden is moved from `winapi` to our crate". By merging this patch as-is, _we_ are now responsible for tracking updates in `windows-rs`. _We_ are now responsible for re-generating the bindings. _We_ are now responsible for publishing a new version with the updated bindings.

This might just be an example, but the general issue extrapolates to other issues such as soundness and security bugs originating in `windows-rs`. Rather than those fixes being _pushed_ to us via crates.io (as is the case with `winapi`), this patch makes it so we need to _pull_ those changes from `windows-rs`. Even if we may not ever update the generated code by hand, the burden for maintaining the bindings still squarely falls on us.

Obviously there's ways around it by setting up infra on the repository. But by doing so we're now on the hook for maintaining that infra, which is no better. I mean it when I say that I'd love for us to move over to `windows-rs`, but both in terms of performance (runtime and compilation), and maintenance overhead it needs to be no worse than `winapi`. Which unfortunately is not true yet today.

Reply:
Sure, there is a bit more work involved here. I see that. The long-term goal is still to support [pre-built versions of APIs](https://github.com/microsoft/windows-rs/issues/432) that would be equivalent in nature to `winapi`. I will just say that getting updates from `winapi` still requires you to pull those updates since you are taking a dependency on a specific version and crates.io will not update that for you automatically. 

At the end of the day, the `Windows` crate is fundamentally different to the `winapi` crate in many ways. It has a completely different philosophy on how API bindings are produced.

Reply:
## Cargo.toml and semver ranges
> I will just say that getting updates from winapi still requires you to pull those updates since you are taking a dependency on a specific version and crates.io will not update that for you automatically.

That's not quite how cargo works treats dependencies. Declaring a dependency in `Cargo.toml` is not pinned to a specific version, instead it specifies a semver range. So imagine I declared the following dependency on `winapi` back in 2017:

```toml
[dependencies]
winapi = "0.3.1"
```

This declares a dependency on the semver range `>=0.3.1, < 0.4.0`, which means that anyone installing our library would automatically receive the latest release of `winapi`, namely `0.3.9`. In the case of `miow` the `winapi` depedency has been declared as `0.3.3` in `Cargo.toml`, but if you look in the `Cargo.lock` of your local checkout you'll find this entry instead:

```toml
[[package]]
name = "winapi"
version = "0.3.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5c839a674fcd7a98952e593242ea400abe93992746761e38641405d28b00f419"
dependencies = [
 "winapi-i686-pc-windows-gnu",
 "winapi-x86_64-pc-windows-gnu",
]
``` 

As with any design, there are tradeoffs. But a distinct benefit of the approach Cargo has taken is that if a transient dependency [^transient] publishes a semver-compatible {bug,security,soundness}-fix, then all of its dependents are automatically upgraded to the latest version.

Pulling in new versions of crates in applications is usually done either on CI by calling `cargo build` in a fresh run. Or sometimes a `Cargo.lock` file is checked in for an application codebase, and `cargo update` will be periodically run to update all dependencies.

[^transient]: if someone takes a dependency on `miow`, then `winapi` is what we call a "transient depencency". AKA "a dependency's dependency".

Reply:
Sure, I get how semver works and I think Rust does a marvelous job with it. But it isn't automatic unless someone initiates a change somewhere along the line. So you are still depending on someone to update your dependencies and that may end up being you if the fix is only important to you as a customer. Looking at `winapi`, it hasn't had an update in over a year and the repo is looking a bit neglected so you may not even be able to get a fix in yourself. My point is just that asserting that the maintenance burden has somehow increased is not necessarily true if the new dependency is more easily or actively maintained and supported. As it is, many projects don't use `winapi` simply because it it lacks thousands of APIs that  the windows crate already provides.

Reply:
By the way, the Windows crate already provides a way to hook into cargo and semver and avoid having to commit and "maintain" the generated code in the same way. You can simply use the `build` macro inside a build script instead of the `generate` macro in a custom tool. That would let cargo automatically update the bindings whenever the windows crate is updated. This works precisely as you've described. The trouble is that this takes a little longer to compile, which is why I opted for the solution presented here. But give it a try - you may find it more appealing. Here's an example:

https://github.com/microsoft/windows-samples-rs/tree/master/overlapped